### DISCUSSION TITLE
JobRunShell silently handles OperationCancelledException

### DISCUSSION THREAD

**Discussion by nstdspace** [quartznet/quartznet#1708] [Q&A]
The `JobRunShell` does nothing besides logging when an OperationCancelledException occurrs, i.e. the job is cancelled via `scheduler.Interrupt`:

```csharp
// Quartz.Core.JobRunShell
catch (OperationCanceledException)
when (jec.CancellationToken.IsCancellationRequested)
{
    endTime = SystemTime.UtcNow();
    logger.LogInformation("Job {JobDetailKey} was cancelled", jobDetail.Key);
}
```

I do not know if that is the whole problem (and if so, I do not understand why) and I did not dug deeper into the code, but the problem is, in `JobListener#JobWasExecuted`, there is no way to find out if `scheduler.Interrupt` was called. The exception is null, and the cancellation token even does not have the correct value for `IsCancellationRequested` (it is always false in the listener).
This goes into the direction of this [SO post](https://stackoverflow.com/questions/68003725/how-to-detect-when-a-job-was-canceled-in-quartz-net), and the workarounds mentioned there are fine, if you are controlling the cancelllation points yourself. Right now, I just catch the exception immediately and throw another one using the following extension method:

```csharp
public static void ThrowIfInterrupted(this IJobExecutionContext context)
{
    try
    {
        context.CancellationToken.ThrowIfCancellationRequested();
    }
    catch (OperationCanceledException exception)
    {
        throw new JobExecutionCanceledException("Job cancelled", exception);
    }
}
```

This works, but does not help if the cancellation is detected in code not controlled by me, for example in an ORM operation.
Also, for whatever reason, in such a scenario, the catch block is not executed, but quartz handles the exception silently:

```csharp
try
{
    await dbContext.DoSomethingAsync(..., cancellationToken);
}
catch (OperationCanceledException exception)
{
    // this does not get calld when a cancellation is detected in the operation above
}
```

What is the reason for that? Is there any reasonable workaround? 
Or, can we maybe just add a flag to the job listener method which indicates if the job was cancelled, or even add another listener method like `OnJobCancelled` ?

Edit: I suspect that the problem could come from the fact that the db operations are awaited async operations. But I'm still not sure what that means exactly in this case.

**Comment by lahma** [comment]
By the looks of it, I think the correct path probably would be to pass the `OperationCanceledException` to the listener *when Quartz knows it was interrupt by looking at the context cancellation token)? Can't remember how Quartz behaved pre-v3 when there was no async yet.

**Reply by nstdspace** [reply]
This would also suffice, I think you have to decide if it is correct ;)
Would you just wrap the exception in a JobExecutionException, then? Because right now this is the type passed to the listener

I would just add the following line, then:

```diff
// JobRunShell line 207
catch (OperationCanceledException exception)
when (jec.CancellationToken.IsCancellationRequested)
{
    endTime = SystemTime.UtcNow();
    log.InfoFormat($"Job {jobDetail.Key} was cancelled");
+   jobExEx = new JobExecutionException(exception)
}
```

**Reply by lahma** [reply]
I checked the v2 codebase and it seems that the silent mode is there also. The change in behavior is that in v2 when you implemented the `Interrup()` in you job level, you could just se a flag. Basically with `while (notInterripted)` logic you job would gracefully exit when interrupt is requested, it's not an exceptional condition.

**Reply by nstdspace** [reply]
I do not claim that it is an exceptional condition, but I think we need a way to detect if a job was cancelled or finished itself.
So how about a method like `OnJobCancelled` in the JobListener?
